arcgis-ios-sdk-samples/Maps/Browse building floors/BrowseBuildingFloorsViewController.swift

```diff
@@ -0,0 +1,115 @@
+        let margin: CGFloat = 10.0
+        floorLevelPickerView.layer.cornerRadius = margin
+        NSLayoutConstraint.activate([
+            floorLevelPickerView.widthAnchor.constraint(equalToConstant: 120.0),
+            floorLevelPickerView.bottomAnchor.constraint(equalTo: mapView.attributionTopAnchor, constant: -margin),
+            floorLevelPickerView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -margin)
+        ])
+    }
+    
+    /// Create a map.
+    func makeMap() -> AGSMap {
+        // A floor-aware web map for floors of Esri Building L in Redlands.
+        let map = AGSMap(item: AGSPortalItem(
+            portal: .arcGISOnline(withLoginRequired: false),
+            itemID: "f133a698536f44c8884ad81f80b6cfc7"
+        ))
+        
+        map.load { [weak self] error in
+            guard error == nil, let self = self,
+                  // The floor manager of the web map, which exposes the sites,
+                  // facilities, and levels of the floor-aware data model.
+                  let floorManager = map.floorManager else { return }
+            
+            floorManager.load { error in
+                guard error == nil, let geometry = floorManager.sites.first?.geometry else { return }
+                self.mapView.setViewpointGeometry(geometry)
+                
+                // Update floor picker and select the first floor.
+                self.levels += floorManager.levels
+                let firstFloorIndex = self.levels.firstIndex { $0?.longName == "Level 1" }!
+                self.setVisibleFloorLevel(self.levels[firstFloorIndex])
+                self.floorLevelPickerView.reloadAllComponents()
+                self.floorLevelPickerView.selectRow(firstFloorIndex, inComponent: 0, animated: true)
+            }
+        }
+        return map
+    }
+    
+    // MARK: UIViewController
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        // Add the source code button item to the right of navigation bar.
+        (navigationItem.rightBarButtonItem as? SourceCodeBarButtonItem)?.filenames = ["BrowseBuildingFloorsViewController"]
+        
+        // Set the appearance of the floor level picker view.
+        setPickerViewLayout()
+    }
+}
+
+// MARK: - UIPickerViewDataSource and UIPickerViewDelegate
+
+extension BrowseBuildingFloorsViewController: UIPickerViewDataSource, UIPickerViewDelegate {
+    func numberOfComponents(in pickerView: UIPickerView) -> Int {
+        1
+    }
+    
+    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
+        levels.count
+    }
+    
+    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
+        levels[row]?.shortName ?? "None"
+    }
+    
+    func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
+        setVisibleFloorLevel(levels[row])
+    }
```

Comment:
Remember how I mentioned not using an array of optionals? Here's how it can be done with an array of non-optionals:

```swift
extension BrowseBuildingFloorsViewController: UIPickerViewDataSource {
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        1
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        floorLevels.count + 1
    }
    
    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        let index = row - 1
        return index >= floorLevels.startIndex ? floorLevels[index].shortName : "None"
    }
}

extension BrowseBuildingFloorsViewController: UIPickerViewDelegate {
    func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        let index = row - 1
        selectedFloorLevel = index >= floorLevels.startIndex ? floorLevels[index] : nil
    }
}
```

And then in the property observer to update the picker view:

```swift
let row: Int
if let selectedFloorLevel = selectedFloorLevel {
    row = floorLevels
        .firstIndex(of: selectedFloorLevel)
        .map { $0 + 1 } ?? .zero
} else {
    row = .zero
}
floorLevelPickerView.selectRow(row, inComponent: 0, animated: true)
```